adding burst class

RobinGuillard opened the pull request:
First step in episode analysis. Defining burst instances with a beggining, an end, the possibility to mark them as tonic and basic functions to manipulate them.

lkorczowski commented:
@RobinGuillard please remove `Notebooks/Bruxism_detection_vbrouillon.ipynb` I don't even understand how it is still there :/

RobinGuillard commented:
Please review the present new branch.
Here are a few explanations of my code:
Starting with a classification result (list or booleans), in order to assess where are the bruxism episodes and what type they are (tonic, phasic or mixed), we need to have a 2 step approach. First we need to "merge" the contiguous bursts (series of True value in the classification result) into single bursts.
This is what the classif_to_burst function does (in the scoring.py code), thanks to the "burst" class.
Then, having in a list all the merged burst, we need to assess where are the real "episodes" of bruxism according to the medical consensus definition. Isolated short bursts should be deleted and only valid episodes that are either tonic, phasic or mixed should remain. For further analysis, it can be interesting though to keep in memory which bursts are contained within each episode.
This is what the burst_to_episode function of scoring does, with the help of the episode class.

At last, we want to be able to compare quickly the scoring of an expert with the automatic scoring, this is why I have made the function create_list_events in scoring.py

Burst and episode classes are here to simplify the scoring.py functions and deal with special cases (overlapping etc...) which will occur rarely.

Surely the code I developed is too much for the issue to handle, I think I can simplify greatly the add_a_burst function tomorrow. Please Take the time to consider that creating the episode labels is not as easy as eit seems before thinking I've done it wrong and starting again from scratch

RobinGuillard commented:
I corrected one error in test_episode, it should work better, and maybe pass the tests!

lkorczowski commented:
`Notebooks/Bruxism_detection_vbrouillon.ipynb → draft/Bruxism_detection_vbrouillon.ipynb`

The gitignore folder is `draft*s*/`

Carefull ^^

RobinGuillard commented:
I made again a pytest trial this morning, and just because I opened up a new terminal, your function "check_if_fitted" began again to throw an error. I'll make an issue out of it. By removing it, it passes all the tests!

RobinGuillard commented:
Ok I really don't know why I don't pass the tests, everything is running fine on my side. and when I click on accessing the codecov report, I get access denied. I'll need you to understand what is happening here ^^

lkorczowski commented:
If you fail codecov coverage, find the non-tested lines by doing in terminal:
```
coverage run -m pytest
coverage report -m
```

the non-covered lines will be mentioned. You could also use PyCharm Professional which has an in-build coverage feature (i.e. non-covered lines are highlighted in your editor)

RobinGuillard commented on tinnsleep/burst.py:
```diff
@@ -0,0 +1,104 @@
+    
+        Returns
+        -------
+        boolean, True if the bursts are overlapping
+        """
+        
+        is_contained= self.beg>=burst.beg and self.end<=burst.end
+        contains= self.beg<=burst.beg and self.end>=burst.end
+        ov_left = (self.beg>=burst.beg and self.end>=burst.end) and (burst.end-self.beg>0)
+        ov_right=(self.beg<=burst.beg and self.end<=burst.end) and (self.end-burst.beg>0)
+        return is_contained or contains or ov_right or ov_left
+            
+    def merge_if_overlap(self, burst):
+        """ test if the current burst is overlapping with another burst and if
+        it is merges the two burst together
+        Parameters
+        ----------
+        self : burst instance
+        burst : burst instance
+    
+        Returns
+        -------
+        boolean, True if the bursts are overlapping
+        """
+        if burst.is_overlapping(self):
+            if self.beg>burst.beg:
+                self.beg = burst.beg
+            if self.end<burst.end:
+                self.end = burst.end
+            
+            if (self.end - self.beg > 2):
+                self.is_tonic=True
+            else:
+                self.is_tonic=False
+            
+            #print(self.beg)
+            #print(self.end)
+            return  True
+        else:
+            return False
```
done :)

RobinGuillard commented on tinnsleep/classification.py:
```diff
@@ -69,7 +69,7 @@ def _predict_distances(self, X):
 
         # Check is fit had been called
-        check_is_fitted(self)
+        #check_is_fitted(self)
```
I deleted the line

RobinGuillard commented on tinnsleep/episode.py:
```diff
@@ -0,0 +1,137 @@
+    """
+
+    def __init__(self, burst):
+        self.beg = burst.beg
+        self.end = burst.end
```
Not sure that episode is directly inheriting from burst as they have different initialisation functions, notably regarding the episode.burst_list attribute. You cannot create an episode the same way you create a burst

RobinGuillard commented on tinnsleep/episode.py:
```diff
@@ -0,0 +1,137 @@
+        elif self.end<=burst.beg:
+            self.end = burst.end
+            self.burst_list.append(burst)
+            self.nb_burst+=1
```
Totally agreeing with you, Suppressing the nb_burst object...

RobinGuillard commented:
I have added some corrections:
- Now burst can be created with end < beg
- Now burst_to_episode works in all cases (thanks to new rearrange_chronological function)
- There was one case of add_a_burst that wasn't covered (Second episode addition with two borns modification). It is now fixed and tested.

RobinGuillard commented on tinnsleep/scoring.py:
```diff
@@ -0,0 +1,143 @@
+from tinnsleep.burst import burst
+from tinnsleep.episode import episode
+
+def classif_to_burst(classif, interval=0.25):
```
Oh yes sorry!

RobinGuillard replied:
But the thing is: if I put an interval in number of samples, then I have to add the s_freq input as well. We need a time interval of 0.25! I keep the time interval but I will name it time_interval for easy differentiation.

RobinGuillard commented:
I still don't get why the coverage tests are not working + despite your explanations, I couldn't run
```
coverage run -m pytest
coverage report -m
```

On my computer. I don't know why it is not working anymore... Another fix to make...

lkorczowski commented on tests/test_episode.py:
```diff
@@ -0,0 +1,201 @@
+def test_episode():
+    bursty = burst(0.1, 1)
+    epi = episode(bursty)
+    npt.assert_equal(epi.beg, 0.1)
+    npt.assert_equal(epi.end, 1)
+    npt.assert_equal(len(epi.burst_list), 1)
```
replace by:
```
npt.assert_equal([epi.beg, epi.end, len(epi.burst_list), [0.1, 1, 1])
```

apply that comment to all multi `npt.assert_equal`

lkorczowski commented on tests/test_episode.py:
```diff
@@ -0,0 +1,201 @@
+    bursty2 = burst(1.75, 3.5)
+    epi = episode(bursty2)
+    epi.add_a_burst(bursty)
+    npt.assert_equal(epi.beg, 1.5)
+    npt.assert_equal(epi.end, 3.5)
+    npt.assert_equal(len(epi.burst_list), 1)
```
multi `npt.assert_equal`

lkorczowski commented on tests/test_episode.py:
```diff
@@ -0,0 +1,201 @@
+    epi.add_a_burst(bursty2)
+    epi.add_a_burst(bursty3)
+    epi.add_a_burst(bursty4)
+    npt.assert_equal(epi.beg, 1.5)
+    npt.assert_equal(epi.end, 5.5)
+    npt.assert_equal(len(epi.burst_list), 2)
```
multi `npt.assert_equal`

lkorczowski commented on tests/test_episode.py:
```diff
@@ -0,0 +1,201 @@
+    bursty2 = burst(2.75, 3.5)
+    epi = episode(bursty2)
+    epi.add_a_burst(bursty)
+    npt.assert_equal(epi.beg, 1.5)
+    npt.assert_equal(epi.end, 3.5)
+    npt.assert_equal(len(epi.burst_list), 2)
```
multi `npt.assert_equal`

lkorczowski commented on tests/test_episode.py:
```diff
@@ -0,0 +1,201 @@
+    bursty2 = burst(1.75, 3.5)
+    epi = episode(bursty)
+    epi.add_a_burst(bursty2)
+    npt.assert_equal(epi.beg, 1.5)
+    npt.assert_equal(epi.end, 3.5)
+    npt.assert_equal(len(epi.burst_list), 1)
```
multi `npt.assert_equal`

lkorczowski commented on tests/test_episode.py:
```diff
@@ -0,0 +1,201 @@
+    epi = episode(bursty)
+    epi.add_a_burst(bursty3)
+    epi.add_a_burst(bursty2)
+    npt.assert_equal(epi.beg, 1.5)
+    npt.assert_equal(epi.end, 3.5)
+    npt.assert_equal(len(epi.burst_list), 2)
```
multi `npt.assert_equal`

lkorczowski commented on tests/test_episode.py:
```diff
@@ -0,0 +1,201 @@
+    epi = episode(bursty)
+    epi.add_a_burst(bursty2)
+    epi.add_a_burst(bursty3)
+    npt.assert_equal(epi.beg, 1.5)
+    npt.assert_equal(epi.end, 3.5)
+    npt.assert_equal(len(epi.burst_list), 1)
```
multi `npt.assert_equal`

lkorczowski commented on tests/test_episode.py:
```diff
@@ -0,0 +1,201 @@
+    bursty2 = burst(2.75, 3.5)
+    epi = episode(bursty)
+    epi.add_a_burst(bursty2)
+    npt.assert_equal(epi.beg, 1.5)
+    npt.assert_equal(epi.end, 3.5)
+    npt.assert_equal(len(epi.burst_list), 2)
```
multi `npt.assert_equal`

lkorczowski commented on tests/test_episode.py:
```diff
@@ -0,0 +1,201 @@
+    epi = episode(bursty)
+    epi.add_a_burst(bursty2)
+    epi.add_a_burst(bursty3)
+    npt.assert_equal(epi.beg, 1.4)
+    npt.assert_equal(epi.end, 3.6)
+    npt.assert_equal(len(epi.burst_list), 1)
```
multi `npt.assert_equal`

lkorczowski commented on tests/test_episode.py:
```diff
@@ -0,0 +1,201 @@
+    bursty2 = burst(1.5, 2.5)
+    epi = episode(bursty2)
+    epi.add_a_burst(bursty)
+    npt.assert_equal(epi.beg, 1.5)
+    npt.assert_equal(epi.end, 2.5)
+    npt.assert_equal(len(epi.burst_list), 1)
```
multi `npt.assert_equal`

lkorczowski commented on tests/test_episode.py:
```diff
@@ -0,0 +1,201 @@
+    epi.add_a_burst(bursty)
+    epi.add_a_burst(bursty3)
+    epi.add_a_burst(bursty4)
+    npt.assert_equal(epi.burst_list[1].beg, 2.7)
+    npt.assert_equal(epi.burst_list[1].end, 3.9)
+    npt.assert_equal(len(epi.burst_list), 3)
```
multi `npt.assert_equal`

lkorczowski commented on tests/test_episode.py:
```diff
@@ -0,0 +1,201 @@
+    epi.add_a_burst(bursty3)
+    epi.add_a_burst(bursty4)
+    epi.add_a_burst(bursty5)
+    npt.assert_equal(epi.burst_list[1].beg, 2.7)
+    npt.assert_equal(epi.burst_list[1].end, 3.9)
+    npt.assert_equal(len(epi.burst_list), 3)
```
multi `npt.assert_equal`

RobinGuillard commented:
Ok It's done!! Moreover, I successfully set up coverage on pycharm, it's working fine :)
